src/main/java/io/github/opencubicchunks/cubicchunks/core/server/VanillaNetworkHandler.java

```diff
@@ -210,50 +255,74 @@ public void sendBlockChanges(TShortList dirtyBlocks, Cube cube, EntityPlayerMP p
                             localToBlock(cube.getZ(), z)));
         }
         ((ISPacketMultiBlockChange) packet).setChangedBlocks(updates);
-        ((ISPacketMultiBlockChange) packet).setChunkPos(cube.getCoords().chunkPos());
+        ((ISPacketMultiBlockChange) packet).setChunkPos(new ChunkPos(posX, posZ));
 
         player.connection.sendPacket(packet);
     }
 
-    public void updatePlayerPosition(PlayerCubeMap cubeMap, EntityPlayerMP player, int managedPosY) {
+    public void updatePlayerPosition(PlayerCubeMap cubeMap, EntityPlayerMP player, CubePos managedPos, int teleportId) {
         if (!CubicChunksConfig.allowVanillaClients) {
             return;
         }
-        if (!playerYOffsets.containsKey(player)) {
-            playerYOffsets.put(player, 0);
-        }
-        int yOffset = playerYOffsets.get(player);
-        if (managedPosY + yOffset < 2 || managedPosY + yOffset >= 14) {
-            int newYOffset = 8 - managedPosY;
-            playerYOffsets.put(player, newYOffset);
-            switchPlayerOffset(cubeMap, player, yOffset, newYOffset);
+        CubePos offset = playerOffsets.get(player);
+        boolean isFirst = offset == null;
+        if (isFirst) {
+            playerOffsets.put(player, offset = CubePos.ZERO);
+        }
+        int posX = managedPos.getX() + offset.getX();
+        int posY = managedPos.getY() + offset.getY();
+        int posZ = managedPos.getZ() + offset.getZ();
+
+        boolean shouldSliceTransition = posY < 2 || posY >= 14;
+        boolean isHorizontalSlices = CubicChunksConfig.vanillaClients.horizontalSlices
+                && (!CubicChunksConfig.vanillaClients.horizontalSlicesBedrockOnly || bedrockPlayers.contains(player.getUniqueID()));
+        if (isHorizontalSlices || teleportId == FORCE_BEDROCK_HORIZONTAL_RECHECK) {
```

Comment:
This constant is only used when a player sends  it's client brand and it's bedrock player. At this moment, the player should already be marked as bedrock player. And if horizontal slices are disabled, this code path seems like it shouldn't be running (and this forces it to run anyway). So why?

Reply:
Correction: this constant is only used when the player sends its client brand, is a Bedrock player, **and** horizontal slices are only enabled for Bedrock players. The client brand is sent later on in the handshake process than the first time that `updatePlayerPosition` is called, which means that if horizontal slices are only enabled for Bedrock players it won't know that the player is on Bedrock at the time and the player will log in at the real, potentially very high, horizontal coordinates, and will then be teleported a second time the moment they send the first move packet. This forces the second teleport to be conducted immediately when the client brand is received, so it's not actually visible to the client.

Reply:
Yes, but when this is called, it's literally a few lines after you marked the player as bedrock. So why is the constant needed? The method call is, I agree. But why this constant to force it, when you literally just added the player as bedrock anyway?

Reply:
Well, marking the player as Bedrock is static, whereas the `VanillaNetworkHandler` instance is a) world-specific and b) not going to be present in a non-cubic world. I guess I could add a separate method called `forceRecheckBedrockPlayerOnAdd`, I just preferred doing it this way to avoid duplicating a ton of code.

Reply:
Yes I do realize that **the call to** `updatePlayerPosition` is necessary here, but **the force recheck flag** shouldn't be. It should just work without it. Is there anything I'm missing there?

From the code that calls it:
```
if (CubicChunksConfig.vanillaClients.horizontalSlices && CubicChunksConfig.vanillaClients.horizontalSlicesBedrockOnly) {
```

And here:
```java
        boolean isHorizontalSlices = CubicChunksConfig.vanillaClients.horizontalSlices
                && (!CubicChunksConfig.vanillaClients.horizontalSlicesBedrockOnly || bedrockPlayers.contains(player.getUniqueID()));
        if (isHorizontalSlices || teleportId == FORCE_BEDROCK_HORIZONTAL_RECHECK) {
```
`CubicChunksConfig.vanillaClients.horizontalSlices` has to be true already when this the force recheck flag is passed.
`CubicChunksConfig.vanillaClients.horizontalSlicesBedrockOnly` is true, so `bedrockPlayers.contains(player.getUniqueID())` is checked. And that check WILL be true because literally a few lines above in `MixinNetHandlerPlayServer` the player is marked as bedrock. This makes `isHorizontalSlices` always true in this case anyway, so the `teleportId == FORCE_BEDROCK_HORIZONTAL_RECHECK` is unnecessary.

Reply:
Apparently github didn't collapse this conversation for me, but dod all the others so i thought the code is still there